[GTK] Added input checking for DateTime.java SWT.TIME Widget

majanojoel opened the pull request:
+ Added validateNewTimeValue() and findNumTimeSeparators() to check for valid key input and gets corrected to a more correct state.

+ Added a check for valid key input for the AM_PM field due to all characters being accepted by this field. Now only alphabetic characters change this field.

Tested using Snippet 250 and ran the DateTime Junit test suite.

Fixes #180 

Signed-off-by: Joel Majano <majanojoel@gmail.com>

majanojoel commented:
Forgot to mention that this bug was creating Gtk-CRITICAL warning messages from GtkSpinButton when the text became longer than 4 characters. Those are not present anymore.

akurtakov commented on bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/DateTime.java:
```diff
@@ -1868,6 +1872,28 @@ boolean onNumberKeyInput(int key) {
+	int numColons = (prefix.length() - prefix.replace(":", "").length()) / ":".length();
+	String finalVal;
+	//numColons mapping: 0 -> hours, 1 -> minutes, 2-> seconds
+	int val = Integer.valueOf(typeBuffer.toString());
```
Please use Integer.parseInt to avoid useless unboxing.

akurtakov replied:
Furthermore this tries to parse AM/FM which  leads into yet another NumberFormatException

majanojoel replied:
Okay, this seems simple enough to fix, run the validtime check only if SWT.TIME && field != AM_PM. I overlooked this point but it is what it should have been in the first place.

akurtakov commented on bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/DateTime.java:
```diff
@@ -1831,6 +1831,9 @@ boolean onNumberKeyInput(int key) {
 		if (!Character.isAlphabetic(newText) && !Character.isDigit(newText)) {
 			return false;
 		}
+		if (fieldName == Calendar.AM_PM && (newText != 'p' && newText !='a')){
```
Upper case A and P do nothing and even thought what if other locales have different indicators for AM/PM ? As per https://en.wikipedia.org/wiki/12-hour_clock#Abbreviations : The equivalents in [Greek](https://en.wikipedia.org/wiki/Greek_language) are π.μ. and μ.μ., respectively, and in [Sinhala](https://en.wikipedia.org/wiki/Sinhala_language) පෙ.ව. (pe.va.) for පෙරවරු (peravaru, පෙර pera – fore, pre) and ප.ව. (pa.va.) for පස්වරු (pasvaru, පස්සේ passē – after, post). So any such filtering should really not happen

majanojoel replied:
Localization support right, okay I'll remove this.

akurtakov commented:
Please redo to only check the numbers to be of the proper size.

majanojoel commented on bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/DateTime.java:
```diff
@@ -1868,6 +1872,28 @@ boolean onNumberKeyInput(int key) {
 }
 
+private StringBuilder validateNewTimeValue(String prefix, String keyValue) {
+	if(typeBuffer.toString() == "") return typeBuffer;
```
For this case, you mention that you were able to delete the hour field, does this mean you were able to delete the colon (":") in the field? If yes, shouldn't this not be allowed. I.e if the key is "Delete" and the charAt the position to be deleted is a colon, this input should be discarded?

majanojoel commented:
> Please redo to only check the numbers to be of the proper size.

I can keep the check for valid time (0-23, 0-59), and the checks for size. For the AM_PM field, due to localization support, I'm not sure of a completely safe way to limit input size, unless it is guaranteed that all locales have the same number of characters in their "A.M"/"P.M" equivalents.

majanojoel commented:
Updated the PR with the required fixes and added a check to only allow non-digit values for the AM_PM field. Let me know if this is still not correct. This still doesn't address the AM_PM field typeBuffer getting too large, but that should be made in a different PR.

mihnita commented on bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/DateTime.java:
```diff
@@ -1868,6 +1875,38 @@ boolean onNumberKeyInput(int key) {
 
+private StringBuilder validateNewTimeValue(String prefix, String keyValue) {
+	if(typeBuffer.toString().isEmpty()) return typeBuffer;
+	int numColons = (prefix.length() - prefix.replace(":", "").length()) / ":".length();
```
Not all locales use ":" as time separator.
Some use "." (be (Belarusian), it (Italian), lt (Lithuanian), sq (Albanian), sr (Serbian))

Worse, this is not 100% stable, might change between Java versions.
The list above is from Java 8.
Java 17 moved to use CLDR by default, so the list of locales using "." grows (with Assamaese, Danish, more)
Worse, separators in 17 go beyond ":" and ".", French-Canada uses " h " ("19 h 43").
And 19:43 in Dzonga (dz) is "ཆུ་ཚོད་ ༧ སྐར་མ་ ༤༣ ཕྱི་ཆ་" (the separator is "སྐར་མ་")

majanojoel replied:
Updated PR to better handle different locales.

majanojoel commented:
I generalized finding the number of separators which seems to work in the en and fr locales (":" and " h" / "min"). I also modified the typing behavior to make it more intuitive. Please let me know if further changes are needed.

akurtakov commented:
Closing as no one seems to be interested in finishing and pushing this one.
